executor/executor.go

```diff
@@ -0,0 +1,605 @@
+}
+
+// SelectLockExec represents a select lock executor.
+type SelectLockExec struct {
+	Src  Executor
+	Lock ast.SelectLockType
+	ctx  context.Context
+}
+
+// Fields implements Executor Fields interface.
+func (e *SelectLockExec) Fields() []*ast.ResultField {
+	return e.Src.Fields()
+}
+
+// Next implements Executor Next interface.
+func (e *SelectLockExec) Next() (*Row, error) {
+	row, err := e.Src.Next()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if row == nil {
+		return nil, nil
+	}
+	if len(row.RowKeys) != 0 && e.Lock == ast.SelectLockForUpdate {
+		forupdate.SetForUpdate(e.ctx)
+		txn, err := e.ctx.GetTxn(false)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, k := range row.RowKeys {
+			err = txn.LockKeys([]byte(k.Key))
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+	}
+	return row, nil
+}
+
+// Close implements Executor Close interface.
+func (e *SelectLockExec) Close() error {
+	return e.Src.Close()
+}
+
+// LimitExec represents limit executor
+type LimitExec struct {
+	Src    Executor
+	Offset uint64
+	Count  uint64
+	Idx    uint64
+}
+
+// Fields implements Executor Fields interface.
+func (e *LimitExec) Fields() []*ast.ResultField {
+	return e.Src.Fields()
+}
+
+// Next implements Executor Next interface.
+func (e *LimitExec) Next() (*Row, error) {
+	for e.Idx < e.Offset {
+		srcRow, err := e.Src.Next()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if srcRow == nil {
+			return nil, nil
+		}
+		e.Idx++
+	}
+	// Negative Limit means no limit.
+	if e.Count >= 0 && e.Idx >= e.Offset+e.Count {
+		return nil, nil
+	}
+	srcRow, err := e.Src.Next()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if srcRow == nil {
+		return nil, nil
+	}
+	e.Idx++
+	return srcRow, nil
+}
+
+// Close implements Executor Close interface.
+func (e *LimitExec) Close() error {
+	return e.Src.Close()
+}
+
+// orderByRow bind a row to its order values, so it can be sorted.
+type orderByRow struct {
+	key []interface{}
+	row *Row
+}
+
+// SortExec represents sorting executor.
+type SortExec struct {
+	Src     Executor
+	ByItems []*ast.ByItem
+	Rows    []*orderByRow
+	ctx     context.Context
+	Idx     int
+	fetched bool
+}
+
+// Fields implements Executor Fields interface.
+func (e *SortExec) Fields() []*ast.ResultField {
+	return e.Src.Fields()
+}
+
+// Len returns the number of rows.
+func (e *SortExec) Len() int {
+	return len(e.Rows)
+}
+
+// Swap implements sort.Interface Swap interface.
+func (e *SortExec) Swap(i, j int) {
+	e.Rows[i], e.Rows[j] = e.Rows[j], e.Rows[i]
+}
+
+// Less implements sort.Interface Less interface.
+func (e *SortExec) Less(i, j int) bool {
+	for index, by := range e.ByItems {
+		v1 := e.Rows[i].key[index]
+		v2 := e.Rows[j].key[index]
+
+		ret, err := types.Compare(v1, v2)
+		if err != nil {
+			// we just have to log this error and skip it.
+			// TODO: record this error and handle it out later.
+			log.Errorf("compare %v %v err %v", v1, v2, err)
```

Comment:
I think now we can use a err variable in SortExec and set it when comparing failed, then we can use Err() function outer to check like database/sql Rows Err().